lib/spack/spack/installer.py

```diff
@@ -0,0 +1,1554 @@
+
+        # The package must have a prefix lock at this stage.
+        if pkg.unique_id not in self.locks:
+            # TODO: If keep, use a custom exception
+            raise spack.error.SpackError('{0} {1}'.format(pre, 'not locked'))
+
+    def _ensure_read_locked(self, pkg):
+        """
+        Add a prefix read lock for the package spec, downgrading from write if
+        needed.
+
+        Args:
+            pkg (spack.package.PackageBase): the package whose spec is being
+                installed
+        """
+        pkg_id = pkg.unique_id
+        ltype, lock = self.locks[pkg_id] if pkg_id in self.locks \
+            else ('read', None)
+        nolock = lock is None
+        if nolock or ltype == 'write':
+            msg = '{0} a read lock on {1} with timeout {2}'
+            err = 'Failed to {0} a read lock for {1} due to {2}: {3}'
+
+            # Wait until the other process finishes if there are no more
+            # build tasks with priority 0 (i.e., with no uninstalled
+            # dependencies).
+            no_p0 = len(self.build_tasks) <= 0 or not self._next_is_pri0()
+            timeout = None if no_p0 else 3
+
+            try:
+                if nolock:
+                    tty.debug(msg.format('Acquiring', pkg_id, timeout))
+                    op = 'acquire'
+                    lock = spack.store.db.prefix_lock(pkg.spec, timeout)
+                    if timeout != lock.default_timeout:
+                        tty.warn('Expected prefix lock timeout {0}, not {1}'
+                                 .format(timeout, lock.default_timeout))
+                    lock.acquire_read()
+                else:
+                    tty.debug(msg.format('Downgrading to', pkg_id,
+                                         lock.default_timeout))
+                    op = 'downgrade to'
+                    # The timeout here defaults to that used to acquire the
+                    # lock.  Since it is most likely a write lock, the timeout
+                    # is near zero.
+
+                    #  TODO: Does providing this resolve console install
+                    #  TODO:   concurrency issues?
+                    # lock.downgrade_write(timeout)
+                    lock.downgrade_write()
+                tty.verbose('{0} is now read locked'.format(pkg_id))
+            except (lk.LockDowngradeError, lk.LockTimeoutError) as exc:
+                tty.debug(err.format(op, pkg_id, exc.__class__.__name__,
+                                     str(exc)))
+                lock = None
+            except (Exception, KeyboardInterrupt, SystemExit) as exc:
+                tty.error(err.format(op, pkg_id, exc.__class__.__name__,
+                          str(exc)))
+                self._cleanup_all_tasks()
+                raise
+
+            self.locks[pkg_id] = ('read', lock)
+
+        return 'read', lock
+
+    def _ensure_write_locked(self, pkg):
```

Comment:
Similar comments as for `_ensure_read_locked`. Do you think we can factor part of the common code together, or the effort is not worth in this case?

Reply:
Refactored into `_ensure_locked`.